src/data/equiv/derangements/basic.lean

```diff
@@ -0,0 +1,218 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.equiv.basic
+import data.equiv.option
+import group_theory.perm.option
+
+import tactic.equiv_rw
+
+/-!
+# Derangements on types
+
+In this file we define a predicate and its corresponding set:
+
+* `is_derangement f`: whether the permutation `f` has no fixed points.
+
+* `derangements α`: the set of derangements on a type `α`.
+
+Then we define
+
+* some related predicates: e.g.,
+
+  - `only_possible_fixed_point f a`: either `f` has no fixed points, or `a` is the only fixed point;
+
+  - `exactly_one_fixed_point f a`: `f` has exactly one fixed point, which is `a`;
+
+* equivalences involving various subtypes of `perm α` and `derangements α`: e.g.,
+
+  - `derangements_equiv_sigma_opfp`: an equivalence between `derangements (option α)` and the
+    sigma-type `Σ a : α, {f : perm α // only_possible_fixed_point f a}`.
+-/
+
+open equiv
+
+namespace perm
+
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | ∀ x : α, f x ≠ x}
+
+/-- The permutation `f` has at most one fixed point, `a`. -/
+def only_possible_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x → x = a
+
+/-- The permutation `f` has exactly one fixed point, `a`. -/
+def exactly_one_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x ↔ x = a
+
+end definitions
+
+section simple_lemmas
+
+/-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
+def derangements_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
+subtype_equiv (perm_congr e) $ λ f, e.forall_congr $ λ x, by simp
+
+end simple_lemmas
+
+section fixed_points
+
+variables {α : Type*}
+
+lemma eofp_iff_opfp_and_eq (a : α) (f : perm α) :
+  exactly_one_fixed_point f a ↔ only_possible_fixed_point f a ∧ f a = a :=
+begin
+  split,
+  { intro h_eofp, exact ⟨λ x, (h_eofp x).mp, (h_eofp a).mpr rfl⟩ },
+  { rintros ⟨h_opfp, fa_eq_a⟩ x,
+    split,
+    { exact h_opfp x },
+    { rintro ⟨rfl⟩, exact fa_eq_a }}
+end
```

Comment:
When is it better to write out the explicit proof term, and when is it better to leave it human readable? Is the former better here because it's pretty mechanically uninteresting?